kafka-persistence/src/main/scala/com/evolutiongaming/kafka/flow/kafkapersistence/KafkaPersistence.scala

```diff
@@ -0,0 +1,167 @@
+
+
+import cats.data.NonEmptyList
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import cats.mtl.MonadState
+import cats.{FlatMap, Monad, data}
+import com.evolutiongaming.catshelper.{BracketThrowable, FromTry, Log}
+import com.evolutiongaming.kafka.flow.KafkaKey
+import com.evolutiongaming.kafka.flow.key.{Keys, KeysOf}
+import com.evolutiongaming.kafka.flow.persistence.{PersistenceOf, SnapshotPersistenceOf}
+import com.evolutiongaming.kafka.flow.snapshot.{SnapshotDatabase, Snapshots, SnapshotsOf}
+import com.evolutiongaming.kafka.journal.util.SkafkaHelper._
+import com.evolutiongaming.skafka.consumer.{ConsumerConfig, ConsumerOf, ConsumerRecord, WithSize}
+import com.evolutiongaming.skafka.producer.{Producer, ProducerRecord}
+import com.evolutiongaming.skafka.{FromBytes, Offset, Partition, ToBytes, Topic, TopicPartition}
+import com.evolutiongaming.sstream.Stream
+import com.olegpy.meow.effects._
+import monocle.macros.GenLens
+import scodec.bits.ByteVector
+
+import scala.concurrent.duration._
+import scala.util.control.NoStackTrace
+
+final case class KafkaPersistence[F[_], K, S, A](
+                                                  keysOf: KeysOf[F, K],
+                                                  snapshots: SnapshotPersistenceOf[F, K, S, A],
+                                                  onRecoveryFinished: F[Unit]
+                                                )
+
+object KafkaPersistence {
+
+  private type Record = ConsumerRecord[String, ByteVector]
+  private type BytesByKey = Map[String, ByteVector]
+
+  private object BytesByKey {
+    val zero: BytesByKey = Map.empty
+  }
+
+  private case class MissingOffsetError(topicPartition: TopicPartition) extends NoStackTrace
+
+  final case class Of[F[_], K, S, A](create: Partition => F[KafkaPersistence[F, K, S, A]])
+
+  object Of {
+    def apply[F[_] : BracketThrowable : Producer : ToBytes[*[_], S] : FromBytes[*[_], S] : FromTry : Log : Sync, S, A](
+                                                                                                                        consumerOf: ConsumerOf[F],
+                                                                                                                        consumerConfig: ConsumerConfig,
+                                                                                                                        stateTopic: Topic
+                                                                                                                      ): Of[F, KafkaKey, S, A] =
+      this { partition =>
+        for {
+          stateData <- readState(
+            consumerOf = consumerOf,
+            consumerConfig = consumerConfig,
+            stateTopic = stateTopic,
+            partition = partition
+          )
+          stateRef <- Ref.of(stateData)
+        } yield stateRef.runState { implicit state =>
+          KafkaPersistence[F, S, A](stateTopic)
```

Comment:
Elsewhere, I pass state explicitly, and then use `stateInstance` method from the same library to pass it. I am not sure which is the best way, but we might want to agree on a single way to do that.

Reply:
Made `monadState` passed explicitly.